language/move-lang/src/expansion/aliases.rs

```diff
@@ -125,78 +123,137 @@ impl AliasMap {
+                    self.unused[*depth].members.remove(n);
+                }
+                *depth_opt = None;
+                Some(ident_name)
+            }
+        }
     }
 
-    pub fn add_and_shadow_all(&mut self, shadowing: AliasMap) {
-        let Self {
+    /// Adds all of the new items in the new inner scope as shadowing the outer one.
+    /// Gives back the outer scope
+    pub fn add_and_shadow_all(&mut self, shadowing: AliasMapBuilder) -> OldAliasMap {
+        if shadowing.is_empty() {
+            return OldAliasMap(None);
+        }
+
+        let outer_scope = OldAliasMap(Some(self.clone()));
+        let AliasMapBuilder {
             modules: new_modules,
             members: new_members,
-            current_scope: new_scope,
         } = shadowing;
-        for (alias, ident) in new_modules {
-            let _ = self.add_implicit_module_alias(alias, ident);
+
+        let next_depth = self.current_depth();
+        let mut current_scope = AliasSet::new();
+        for (alias, (ident, is_implicit)) in new_modules {
+            if !is_implicit {
+                current_scope.modules.add(alias.clone()).unwrap();
+            }
+            self.modules.remove(&alias);
+            self.modules.add(alias, (Some(next_depth), ident)).unwrap();
+        }
+        for (alias, (ident_member, is_implicit)) in new_members {
+            if !is_implicit {
+                current_scope.members.add(alias.clone()).unwrap();
+            }
+            self.members.remove(&alias);
+            self.members
+                .add(alias, (Some(next_depth), ident_member))
+                .unwrap();
+        }
+        self.unused.push(current_scope);
+        outer_scope
+    }
+
+    /// Similar to add_and_shadow but just removes aliases now shadowed by a type parameter
+    pub fn shadow_for_type_parameters<'a, I: IntoIterator<Item = &'a Name>>(
+        &mut self,
+        tparams: I,
+    ) -> OldAliasMap
+    where
+        I::IntoIter: ExactSizeIterator,
+    {
+        let tparams_iter = tparams.into_iter();
+        if tparams_iter.len() == 0 {
+            return OldAliasMap(None);
         }
-        for (alias, (ident, member)) in new_members {
-            let _ = self.add_implicit_member_alias(alias, ident, member);
+
+        let outer_scope = OldAliasMap(Some(self.clone()));
+        self.unused.push(AliasSet::new());
+        for tp_name in tparams_iter {
+            self.members.remove(tp_name);
         }
-        self.current_scope = new_scope
+        outer_scope
     }
 
-    pub fn close_scope_and_report_unused(&mut self, inner: Self) -> AliasSet {
+    /// Resets the alias map and gives the set of aliases that were used
+    pub fn set_to_outer_scope(&mut self, outer_scope: OldAliasMap) -> AliasSet {
+        let outer_scope = match outer_scope.0 {
+            None => return AliasSet::new(),
+            Some(outer) => outer,
+        };
+        let mut inner_scope = std::mem::replace(self, outer_scope);
```

Comment:
Wow!! TIL you can do something like this in Rust.

Reply:
std::mem is a fun place :)